engineering-decision-records/007-observability.md

```diff
@@ -0,0 +1,52 @@
+
+Each module or crate will, if applicable, accept an argument of type `{CrateName}Metrics` during initialization, and store an owned instance of this struct.
+
+Each metrics struct will expose a number of top-level [Counters, Gauges, or Histograms](https://docs.rs/opentelemetry/latest/opentelemetry/metrics/index.html) relevant to its workload, and in the course of performing its work, will update these accordingly. The `new(..)` method will accept an [`SdkMeterProvider`](https://docs.rs/opentelemetry_sdk/latest/opentelemetry_sdk/metrics/struct.SdkMeterProvider.html) and initialize each of the relevant objects with names, descriptions, and units as appropriate.
+
+- [Counters](https://opentelemetry.io/docs/specs/otel/metrics/api/#counter) are used to count discrete or measurable events or quantities that can be accumulated over time, such as number of blocks processed, number of bytes read, etc.
+- [Guages](https://opentelemetry.io/docs/specs/otel/metrics/api/#gauge) are used to track measurements that only make sense at a point in time, and shouldn't be added, such as current Memory or CPU usage, temperature, etc.
+- [Histograms](https://opentelemetry.io/docs/specs/otel/metrics/api/#histogram) are used to track values where the statistical distributions are important, such as response latencies, where you want to query the 90th percentile, etc.
+
+Such metrics should follow the [Open Telemetry Metrics Semantics conventions](https://opentelemetry.io/docs/specs/semconv/general/metrics/).
+
+The Amaru binary itself will contain a `metrics` module that stores and constructs instances of all other metrics types. It will be responsible for constructing and orchestrating all of the relevant modules and crates. It will also track common system and process metrics, such as CPU and memory usage.
+
+## Consequence
+
+As a consequence, each component of the Amaru node can fully own its own metrics. At the same time, someone wishing to document or explore the metrics supported by the Amaru node has a single entrypoint to begin their exploration.
+
+Users of the Amaru node can track, aggregate, and alert on the health of the node based on a variety of system or component specific metrics. They might notice, for example, that CPU usage hits 100% at epoch boundaries, resulting in missing slot leader checks, which indicates that the machine they are running the node on is underprovisioned.
+
+## Discussion points
+
+One thing we discussed was whether to use the notion of Gasket metrics or not. Gasket is a pipeline processing library written by one of the maintainers of Amaru, and provides some light metrics tracking machinery that also exposes things via prometheus. At this time we've decided not to use these metrics, as it's not the primary responsibility of the Gasket framework, so that the Amaru node is in full control of the metrics it tracks and reports.
```

Comment:
This comment could be made generic and mention any libraries, not just `gasket`.